### ISSUE TITLE
cmake dependencies to opencv could be partial

### ISSUE THREAD

**Issue by tkruse** [ros-perception/image_pipeline#37]
I read some blogpost about overlinking and .pc files, and then wondered how ROS currently fares with overlinking. (E.g. see http://wiki.mandriva.com/en/Overlinking)
So I did a grep over .pc files in a groovy desktop full underlay, and the things that stood out where opencv dependencies.

E.g.:

```
$ find devel_isolated/ -name *.pc | xargs grep Libs
devel_isolated/image_geometry/lib/pkgconfig/image_geometry.pc:Libs: -L/home/kruset/groovy_underlay_full/devel_isolated/image_geometry/lib -limage_geometry -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_calib3d.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_contrib.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_core.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_features2d.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_flann.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_gpu.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_highgui.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_imgproc.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_legacy.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_ml.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_nonfree.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_objdetect.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_photo.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_stitching.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_superres.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_video.so -l:/home/kruset/groovy_underlay_full/devel_isolated/opencv2/lib/libopencv_videostab.so
```

Same for cv_bridge, compressed_image_transport, compressed_depth_image_transport, image_proc

I assume that linking here is overdone, and could / should be reduced by getting opencv into cmake with components, e.g.:
http://stackoverflow.com/questions/11196720/cmake-how-to-only-include-part-of-opencv

```
FIND_PACKAGE( OpenCV REQUIRED core imgproc)
```

I don't know much about opencv, cmake or linking, so I might be wrong here, but I still thought it might be worth raising a flag.

**Comment by vrabaud** [comment]
If you do an ldd on your libraries, they are only linked to the appropriate libraries, nothing more.
So what you would get with FIND_PACKAGE( OpenCV REQUIRED core imgproc) is just a slightly faster compilation time. Is there anything else you would expect to get ?

**Comment by tkruse** [comment]
No, just some cleanliness maybe. I have no larger worry about this.

**Comment by vrabaud** [comment]
ok, so your point is actually different: it's about having the catkin_package macro not being able to use modules right ? Actually, I think it does, but it is not documented. Or at least we talked about it with @dirk-thomas

**Comment by tkruse** [comment]
No, for me it would be fine to have
FIND_PACKAGE( OpenCV REQUIRED core imgproc ...)
instead of
FIND_PACKAGE( OpenCV)   # the current CmakeLists.txt content e.g. for
image_proc

That way, it is self-documenting what parts of opencv this uses (and which
not), and also you can quickly see for a pull request whether the author
has used an openCV function that he should not have used (modification of
CMakeLists.txt. It also is good practice because people copy and paste from
other CMakeLists.txt, so any dirty CMakeLists.txt spawns more dirty
projects.

Maybe there is something I don't understand here, but i see no connection
to the catkin_package macro.

**Comment by vrabaud** [comment]
As I mentioned earlier, the generated libraries are not overlinked: they are told to be linked with all of OpenCV but the compiler is smart enough to remove the unused ones. So doing your FIND_PACKAGE( OpenCV REQUIRED core imgproc ...) is cleaner, but the end result is the same for the libraries.

Now, the problem you mentioned in your first comment is related to the *.pc files, that are generated by catkin. In those, there are all the libraries of OpenCV. This is indeed bad because it forces all the dependents to try to link to all of OpenCV (which is not much CPU usage but still, can be a lot when aggregating over time). Now, if FIND_PACKAGE( OpenCV REQUIRED core imgproc ...) fixes that, great, that is the solution. My understanding, and I might be wrong, is that it does not. A solution could be to have catkin_package understand something like:
catkin_package(DEPENDS OpenCV REQUIRED core imgproce DEPENDS Boost ...)
This way the *pc would be clean.

**Comment by tkruse** [comment]
You are right, those are two different problems, I did not see that before.
Using the right components in find_package would be a bit cleaner, and
would also reduce the items in the .pc file. (Having less flags generally
can also help debugging a bit)

However, to be maximally clean, in catkin_package we'd have to further
reduce the exported components.

BTW: I also noticed in the process that in stereo_image_proc, the
catkin_package macro seems to not export dependency, though it seems to
have a public header file.
